File: src/domain/architecture-model/references.mjs
import { releaseSummaryFromDetail } from "./release-history.mjs";
import { releaseItems } from "./release-scopes.mjs";

export function validateArchitectureReferences(model) {
  const errors = [];
  const arrayField = (owner, field, required = false) => {
    const value = owner?.[field];
    if (Array.isArray(value)) return value;
    if (value === undefined && !required) return [];
    errors.push(`${field} must be an array`);
    return [];
  };
  const nestedArray = (owner, field, context) => {
    const value = owner?.[field];
    if (Array.isArray(value)) return value;
    if (value === undefined) return [];
    errors.push(`${context}.${field} must be an array`);
    return [];
  };
  const nodes = arrayField(model, "nodes", true);
  const flows = arrayField(model, "flows", true);
  const dataClasses = arrayField(model, "dataClasses", true);
  const decisions = arrayField(model, "decisions", true);
  const risks = arrayField(model, "risks", true);
  const views = arrayField(model, "views", true);
  const nodeIds = new Set(nodes.map((node) => node.id));
  const flowIds = new Set(flows.map((flow) => flow.id));
  const dataIds = new Set(dataClasses.map((item) => item.id));
  const decisionIds = new Set(decisions.map((item) => item.id));
  const riskIds = new Set(risks.map((item) => item.id));
  const viewIds = new Set(views.map((item) => item.id));

  const requireKnown = (id, known, context) => {
    if (!known.has(id)) errors.push(`${context} references unknown id "${id}"`);
  };

  requireKnown(model.manifest.defaultViewId, viewIds, "manifest.defaultViewId");

  for (const node of nodes) {
    for (const id of nestedArray(node, "dependencies", `node ${node.id}`)) requireKnown(id, nodeIds, `node ${node.id}.dependencies`);
    for (const id of nestedArray(node, "dataHandled", `node ${node.id}`)) requireKnown(id, dataIds, `node ${node.id}.dataHandled`);
    for (const id of nestedArray(node, "relatedFlows", `node ${node.id}`)) requireKnown(id, flowIds, `node ${node.id}.relatedFlows`);
    for (const id of nestedArray(node, "relatedDecisions", `node ${node.id}`)) requireKnown(id, decisionIds, `node ${node.id}.relatedDecisions`);
    for (const id of nestedArray(node, "knownRisks", `node ${node.id}`)) requireKnown(id, riskIds, `node ${node.id}.knownRisks`);
  }

  for (const flow of flows) {
    const steps = nestedArray(flow, "steps", `flow ${flow.id}`);
    const stepIds = new Set(steps.map((step) => step.id));
    for (const id of nestedArray(flow, "actors", `flow ${flow.id}`)) requireKnown(id, nodeIds, `flow ${flow.id}.actors`);
    for (const step of steps) {
      if (!step.from) errors.push(`flow ${flow.id} step ${step.id}.from is required`);
      else requireKnown(step.from, nodeIds, `flow ${flow.id} step ${step.id}.from`);
      if (!step.to) errors.push(`flow ${flow.id} step ${step.id}.to is required`);
      else requireKnown(step.to, nodeIds, `flow ${flow.id} step ${step.id}.to`);
      if (step.returnOf) requireKnown(step.returnOf, stepIds, `flow ${flow.id} step ${step.id}.returnOf`);
      for (const id of nestedArray(step, "data", `flow ${flow.id} step ${step.id}`)) requireKnown(id, dataIds, `flow ${flow.id} step ${step.id}.data`);
    }
    for (const frame of nestedArray(flow, "sequenceFrames", `flow ${flow.id}`)) {
      for (const id of nestedArray(frame, "stepIds", `flow ${flow.id} sequenceFrame ${frame.id}`)) requireKnown(id, stepIds, `flow ${flow.id} sequenceFrame ${frame.id}.stepIds`);
    }
  }

  for (const view of views) {
    if (view.scopeNodeId) requireKnown(view.scopeNodeId, nodeIds, `view ${view.id}.scopeNodeId`);
    for (const lane of nestedArray(view, "lanes", `view ${view.id}`)) {
      for (const id of nestedArray(lane, "nodeIds", `view ${view.id} lane ${lane.id}`)) requireKnown(id, nodeIds, `view ${view.id} lane ${lane.id}`);
    }
  }

  if (Array.isArray(model.notes)) {
    const idsByKind = {
      node: nodeIds, flow: flowIds, decision: decisionIds,
      risk: riskIds, view: viewIds, "data-class": dataIds
    };
    for (const note of model.notes) {
      const known = idsByKind[note.target?.kind];
      if (!known) errors.push(`note ${note.id}.target.kind "${note.target?.kind}" is unknown`);
      else requireKnown(note.target.id, known, `note ${note.id}.target (${note.target.kind})`);
    }
  }

  if (model.releases) {
    validateReleaseReferences(model.releases, errors, { requireAllDetails: false });
  }
  if (model.roadmap && model.releases) {
    const releaseIds = new Set(model.releases.index.releases.map((release) => release.id));
    for (const item of model.roadmap) {
      if (item.targetReleaseId) requireKnown(item.targetReleaseId, releaseIds, `roadmap item ${item.id}.targetReleaseId`);
    }
  }

  return errors;
}

function releaseItemCanBeBlocked(status) {
  return !["complete", "deferred", "cut"].includes(status);
}

export function validateReleaseReferences(releases, errors = [], options = {}) {
  const requireAllDetails = options.requireAllDetails ?? true;
  const releaseIds = new Set(releases.index.releases.map((release) => release.id));
  const detailIds = new Set(releases.details.map((detail) => detail.id));

  const requireKnown = (id, known, context) => {
    if (!known.has(id)) errors.push(`${context} references unknown id "${id}"`);
  };

  requireKnown(releases.index.currentReleaseId, releaseIds, "releases.currentReleaseId");

  for (const summary of releases.index.releases) {
    if (requireAllDetails) requireKnown(summary.id, detailIds, `release index ${summary.id}.file`);
    const detail = releases.details.find((item) => item.id === summary.id);
    if (!detail) continue;
    if (detail.version !== summary.version) {
      errors.push(`release ${summary.id}.version does not match release index`);
    }
    if (detail.status !== summary.status) {
      errors.push(`release ${summary.id}.status does not match release index`);
    }
    if (!sameGeneratedReleaseSummary(summary, detail)) {
      errors.push(`release ${summary.id}.index summary is stale; regenerate Release Truth history`);
    }
    if (summary.status === "completed" && !summary.releasedAt) {
      errors.push(`release index ${summary.id}.releasedAt is required for completed entries`);
    }
    if ((summary.status === "implementing" || summary.status === "planned" || summary.status === "draft") && !summary.targetDate && !summary.targetWindow) {
      errors.push(`release index ${summary.id} requires targetDate or targetWindow`);
    }
  }

  for (const detail of releases.details) {
    const items = releaseItems(detail);
    const itemIds = new Set(items.map((item) => item.id));
    const itemsById = new Map(items.map((item) => [item.id, item]));
    const workstreamIds = new Set(detail.workstreams.map((workstream) => workstream.id));

    if (detail.status === "completed" && !detail.releasedAt) {
      errors.push(`release ${detail.id}.releasedAt is required for completed entries`);
    }
    if ((detail.status === "implementing" || detail.status === "planned" || detail.status === "draft") && !detail.targetDate && !detail.targetWindow) {
      errors.push(`release ${detail.id} requires targetDate or targetWindow`);
    }

    for (const item of items) {
      if (item.workstreamId) requireKnown(item.workstreamId, workstreamIds, `release ${detail.id} item ${item.id}.workstreamId`);
      for (const dependencyId of item.dependsOn ?? []) requireKnown(dependencyId, itemIds, `release ${detail.id} item ${item.id}.dependsOn`);
    }

    for (const workstream of detail.workstreams) {
      for (const itemId of workstream.itemIds) requireKnown(itemId, itemIds, `release ${detail.id} workstream ${workstream.id}.itemIds`);
    }

    for (const blocker of detail.blockers) {
      for (const itemId of blocker.itemIds) {
        requireKnown(itemId, itemIds, `release ${detail.id} blocker ${blocker.id}.itemIds`);
        const item = itemsById.get(itemId);
        if (item && !releaseItemCanBeBlocked(item.status)) {
          errors.push(`release ${detail.id} blocker ${blocker.id}.itemIds references ${item.status} item "${itemId}"`);
        }
      }
    }

    for (const milestone of detail.milestones) {
      for (const itemId of milestone.itemIds) requireKnown(itemId, itemIds, `release ${detail.id} milestone ${milestone.id}.itemIds`);
    }

    for (const dependency of detail.dependencies) {
      requireKnown(dependency.from, itemIds, `release ${detail.id} dependency ${dependency.id}.from`);
      requireKnown(dependency.to, itemIds, `release ${detail.id} dependency ${dependency.id}.to`);
    }
  }

  return errors;
}

function sameGeneratedReleaseSummary(summary, detail) {
  const generated = releaseSummaryFromDetail(detail, summary.file);
  return JSON.stringify(normalizeReleaseSummary(summary)) === JSON.stringify(normalizeReleaseSummary(generated));
}

function normalizeReleaseSummary(summary) {
  return {
    id: summary.id,
    version: summary.version,
    name: summary.name,
    status: summary.status,
    posture: summary.posture,
    targetDate: summary.targetDate ?? null,
    targetWindow: summary.targetWindow ?? null,
    releasedAt: summary.releasedAt ?? null,
    lastUpdated: summary.lastUpdated,
    summary: summary.summary,
    counts: summary.counts,
    file: summary.file
  };
}

File: crates/architext-core/src/validation/references.rs
//! Reference-integrity validation layer.
//!
//! Mirrors the `requireUnique` and `validateReferences` functions in
//! `viewer/tools/validate-architext.mjs` (the inline ones called by `main()`).
//! Error strings are reproduced verbatim so the product's deep-validate
//! comparisons continue to work.

use std::collections::{HashMap, HashSet};
use std::path::Path;

use serde_json::Value;

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/// Push an error if `id` is not in `known`.
/// Mirrors: `function requireKnown(id, known, context, errors)`
fn require_known(id: &str, known: &HashSet<&str>, context: &str, errors: &mut Vec<String>) {
    if !known.contains(id) {
        errors.push(format!("{context} references unknown id \"{id}\""));
    }
}

/// Push an error for every duplicate `id` field among `items`.
/// Mirrors: `function requireUnique(items, label, errors)`
fn require_unique<'a>(
    items: impl Iterator<Item = &'a Value>,
    label: &str,
    errors: &mut Vec<String>,
) {
    let mut seen: HashSet<&str> = HashSet::new();
    for item in items {
        let id = match item.get("id").and_then(Value::as_str) {
            Some(s) => s,
            None => continue,
        };
        if seen.contains(id) {
            errors.push(format!("{label} contains duplicate id \"{id}\""));
        }
        seen.insert(id);
    }
}

/// Collect all `id` string values from an array of objects into a `HashSet`.
fn id_set(items: &[Value]) -> HashSet<&str> {
    items
        .iter()
        .filter_map(|v| v.get("id").and_then(Value::as_str))
        .collect()
}

/// Read a JSON file; return `None` and push an error on failure.
fn read_json(path: &Path, errors: &mut Vec<String>) -> Option<Value> {
    let text = match std::fs::read_to_string(path) {
        Ok(t) => t,
        Err(e) => {
            errors.push(format!("cannot read {}: {}", path.display(), e));
            return None;
        }
    };
    match serde_json::from_str::<Value>(&text) {
        Ok(v) => Some(v),
        Err(e) => {
            errors.push(format!("invalid JSON in {}: {}", path.display(), e));
            None
        }
    }
}

// ---------------------------------------------------------------------------
// Public entry point
// ---------------------------------------------------------------------------

/// Run duplicate-id and cross-file reference checks on the data directory.
///
/// Mirrors `requireUnique(…)` calls and `validateReferences(model, errors)`
/// from `viewer/tools/validate-architext.mjs main()`.
///
/// Called unconditionally after schema validation (JS does the same —
/// `validateReferences` is called even when schema errors were pushed).
pub fn validate_references(data_dir: &Path, errors: &mut Vec<String>) {
    // --- load manifest -------------------------------------------------------
    let manifest_path = data_dir.join("manifest.json");
    let manifest = match read_json(&manifest_path, errors) {
        Some(v) => v,
        None => return, // cannot continue
    };

    let files = match manifest.get("files").and_then(Value::as_object) {
        Some(f) => f,
        None => return, // schema layer already errors on this
    };

    // Helper: load a required file by manifest key.
    macro_rules! load_required {
        ($key:expr) => {{
            let rel = match files.get($key).and_then(Value::as_str) {
                Some(s) => s,
                None => return, // schema layer already reported this
            };
            match read_json(&data_dir.join(rel), errors) {
                Some(v) => v,
                None => return,
            }
        }};
    }

    // Helper: load an optional file by manifest key (None if absent).
    macro_rules! load_optional {
        ($key:expr) => {{
            match files.get($key).and_then(Value::as_str) {
                Some(rel) => read_json(&data_dir.join(rel), errors),
                None => None,
            }
        }};
    }

    let nodes_doc = load_required!("nodes");
    let flows_doc = load_required!("flows");
    let views_doc = load_required!("views");
    let data_doc = load_required!("dataClassification");
    let decisions_doc = load_required!("decisions");
    let risks_doc = load_required!("risks");
    let notes_doc = load_optional!("notes");
    let roadmap_doc = load_optional!("roadmap");
    let rules_doc = load_optional!("rules");

    // --- extract arrays ------------------------------------------------------
    let empty = vec![];

    let nodes = nodes_doc.get("nodes").and_then(Value::as_array).unwrap_or(&empty);
    let flows = flows_doc.get("flows").and_then(Value::as_array).unwrap_or(&empty);
    let views = views_doc.get("views").and_then(Value::as_array).unwrap_or(&empty);
    let data_classes = data_doc.get("classes").and_then(Value::as_array).unwrap_or(&empty);
    let decisions = decisions_doc.get("decisions").and_then(Value::as_array).unwrap_or(&empty);
    let risks = risks_doc.get("risks").and_then(Value::as_array).unwrap_or(&empty);

    // --- requireUnique (mirrors JS main() call order) ------------------------
    require_unique(nodes.iter(), "nodes", errors);
    require_unique(flows.iter(), "flows", errors);
    require_unique(views.iter(), "views", errors);
    require_unique(data_classes.iter(), "dataClassification.classes", errors);
    require_unique(decisions.iter(), "decisions", errors);
    require_unique(risks.iter(), "risks", errors);
    if let Some(ref rm) = roadmap_doc {
        let items = rm.get("items").and_then(Value::as_array).unwrap_or(&empty);
        require_unique(items.iter(), "roadmap.items", errors);
    }
    if let Some(ref rl) = rules_doc {
        let rules_arr = rl.get("rules").and_then(Value::as_array).unwrap_or(&empty);
        require_unique(rules_arr.iter(), "rules", errors);
    }
    if let Some(ref nd) = notes_doc {
        let notes_arr = nd.get("notes").and_then(Value::as_array).unwrap_or(&empty);
        require_unique(notes_arr.iter(), "notes", errors);
    }

    // --- build id sets -------------------------------------------------------
    let node_ids: HashSet<&str> = id_set(nodes);
    let flow_ids: HashSet<&str> = id_set(flows);
    let data_ids: HashSet<&str> = id_set(data_classes);
    let decision_ids: HashSet<&str> = id_set(decisions);
    let risk_ids: HashSet<&str> = id_set(risks);
    let view_ids: HashSet<&str> = id_set(views);

    // --- manifest.defaultViewId ----------------------------------------------
    if let Some(default_view_id) = manifest.get("defaultViewId").and_then(Value::as_str) {
        require_known(default_view_id, &view_ids, "manifest.defaultViewId", errors);
    }

    // --- nodes ---------------------------------------------------------------
    for node in nodes {
        let node_id = match node.get("id").and_then(Value::as_str) {
            Some(s) => s,
            None => continue,
        };
        for id in arr_strs(node, "dependencies") {
            require_known(id, &node_ids, &format!("node {node_id}.dependencies"), errors);
        }
        for id in arr_strs(node, "dataHandled") {
            require_known(id, &data_ids, &format!("node {node_id}.dataHandled"), errors);
        }
        for id in arr_strs(node, "relatedFlows") {
            require_known(id, &flow_ids, &format!("node {node_id}.relatedFlows"), errors);
        }
        for id in arr_strs(node, "relatedDecisions") {
            require_known(id, &decision_ids, &format!("node {node_id}.relatedDecisions"), errors);
        }
        for id in arr_strs(node, "knownRisks") {
            require_known(id, &risk_ids, &format!("node {node_id}.knownRisks"), errors);
        }
    }

    // --- flows ---------------------------------------------------------------
    for flow in flows {
        let flow_id = match flow.get("id").and_then(Value::as_str) {
            Some(s) => s,
            None => continue,
        };
        for id in arr_strs(flow, "actors") {
            require_known(id, &node_ids, &format!("flow {flow_id}.actors"), errors);
        }

        let steps = flow.get("steps").and_then(Value::as_array).unwrap_or(&empty);

        // Build step id set (with duplicate detection) — mirrors JS two-pass.
        let mut step_ids: HashSet<&str> = HashSet::new();
        for step in steps.iter() {
            let step_id = match step.get("id").and_then(Value::as_str) {
                Some(s) => s,
                None => continue,
            };
            if step_ids.contains(step_id) {
                errors.push(format!(
                    "flow {flow_id} contains duplicate step id \"{step_id}\""
                ));
            }
            step_ids.insert(step_id);
        }

        // Validate step references.
        for step in steps.iter() {
            let step_id = match step.get("id").and_then(Value::as_str) {
                Some(s) => s,
                None => continue,
            };
            if let Some(from) = step.get("from").and_then(Value::as_str) {
                require_known(from, &node_ids, &format!("flow {flow_id} step {step_id}.from"), errors);
            }
            if let Some(to) = step.get("to").and_then(Value::as_str) {
                require_known(to, &node_ids, &format!("flow {flow_id} step {step_id}.to"), errors);
            }
            if let Some(return_of) = step.get("returnOf").and_then(Value::as_str) {
                require_known(return_of, &step_ids, &format!("flow {flow_id} step {step_id}.returnOf"), errors);
            }
            for id in arr_strs(step, "data") {
                require_known(id, &data_ids, &format!("flow {flow_id} step {step_id}.data"), errors);
            }
        }

        // sequenceFrames
        let frames = flow.get("sequenceFrames").and_then(Value::as_array);
        if let Some(frames) = frames {
            for frame in frames {
                let frame_id = match frame.get("id").and_then(Value::as_str) {
                    Some(s) => s,
                    None => continue,
                };
                for id in arr_strs(frame, "stepIds") {
                    require_known(
                        id,
                        &step_ids,
                        &format!("flow {flow_id} sequenceFrame {frame_id}.stepIds"),
                        errors,
                    );
                }
            }
        }
    }

    // --- views ---------------------------------------------------------------
    for view in views {
        let view_id = match view.get("id").and_then(Value::as_str) {
            Some(s) => s,
            None => continue,
        };
        let lanes = view.get("lanes").and_then(Value::as_array).unwrap_or(&empty);
        let mut lane_ids: HashSet<&str> = HashSet::new();
        for lane in lanes {
            let lane_id = match lane.get("id").and_then(Value::as_str) {
                Some(s) => s,
                None => continue,
            };
            if lane_ids.contains(lane_id) {
                errors.push(format!(
                    "view {view_id} contains duplicate lane id \"{lane_id}\""
                ));
            }
            lane_ids.insert(lane_id);
            for id in arr_strs(lane, "nodeIds") {
                require_known(id, &node_ids, &format!("view {view_id} lane {lane_id}"), errors);
            }
        }
    }

    // --- decisions -----------------------------------------------------------
    for decision in decisions {
        let decision_id = match decision.get("id").and_then(Value::as_str) {
            Some(s) => s,
            None => continue,
        };
        for id in arr_strs(decision, "relatedNodes") {
            require_known(id, &node_ids, &format!("decision {decision_id}.relatedNodes"), errors);
        }
        for id in arr_strs(decision, "relatedFlows") {
            require_known(id, &flow_ids, &format!("decision {decision_id}.relatedFlows"), errors);
        }
    }

    // --- risks ---------------------------------------------------------------
    for risk in risks {
        let risk_id = match risk.get("id").and_then(Value::as_str) {
            Some(s) => s,
            None => continue,
        };
        for id in arr_strs(risk, "relatedNodes") {
            require_known(id, &node_ids, &format!("risk {risk_id}.relatedNodes"), errors);
        }
        for id in arr_strs(risk, "relatedFlows") {
            require_known(id, &flow_ids, &format!("risk {risk_id}.relatedFlows"), errors);
        }
    }

    // --- notes ---------------------------------------------------------------
    if let Some(ref nd) = notes_doc {
        let notes_arr = nd.get("notes").and_then(Value::as_array).unwrap_or(&empty);
        // Build the kind → id-set map.
        let ids_by_kind: HashMap<&str, &HashSet<&str>> = [
            ("node", &node_ids),
            ("flow", &flow_ids),
            ("decision", &decision_ids),
            ("risk", &risk_ids),
            ("view", &view_ids),
            ("data-class", &data_ids),
        ]
        .into_iter()
        .collect();

        for note in notes_arr {
            let note_id = match note.get("id").and_then(Value::as_str) {
                Some(s) => s,
                None => continue,
            };
            let target = match note.get("target") {
                Some(t) => t,
                None => continue,
            };
            let kind = match target.get("kind").and_then(Value::as_str) {
                Some(s) => s,
                None => continue,
            };
            let target_id = match target.get("id").and_then(Value::as_str) {
                Some(s) => s,
                None => continue,
            };
            if let Some(known) = ids_by_kind.get(kind) {
                require_known(
                    target_id,
                    known,
                    &format!("note {note_id}.target ({kind})"),
                    errors,
                );
            }
        }
    }
}

// ---------------------------------------------------------------------------
// Internal utility
// ---------------------------------------------------------------------------

/// Iterate over string values in a JSON array field of an object.
fn arr_strs<'a>(obj: &'a Value, field: &str) -> impl Iterator<Item = &'a str> {
    obj.get(field)
        .and_then(Value::as_array)
        .map(|arr| arr.as_slice())
        .unwrap_or(&[])
        .iter()
        .filter_map(|v| v.as_str())
}
